Add `FieldLevelHelp` component

AparnaKarve opened the pull request:
Added a new component called `FieldLevelHelp` based on the design requirements in  http://www.patternfly.org/pattern-library/forms-and-controls/help-on-forms/

[FieldLevelHelp Storybook](https://rawgit.com/AparnaKarve/patternfly-react/add_fieldlevelhelp_component-storybook/index.html?knob-Show%20Modal=true&knob-content=Enter%20the%20hostname%20in%20a%20valid%20format%20%3Cbr%3E%20%20%3Ca%20target%3D%22_blank%22%20href%3D%22http%3A%2F%2Fwww.test.example.com%22%3EClick%20here%20for%20examples%20of%20valid%20hostnames%3C%2Fa%3E&knob-Close%20Popover=true&knob-Show%20Labels=false&knob-Show%20Help=true&knob-Show%20Loading=false&knob-Disabled=false&knob-Field%20Label=Hostname&knob-FieldLevelHelp%20Content=Please%20specify%20Country%20code%20%3Cbr%3E%20%3Ca%20target%3D%27_blank%27%20href%3D%27https%3A%2F%2Fcountrycode.org%2F%27%3EClick%20here%20for%20a%20list%20of%20Country%20codes%3C%2Fa%3E&selectedKind=Forms&selectedStory=Inline%20Form&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

Fixes https://github.com/patternfly/patternfly-react/issues/115

serenamarie125 commented:
@AparnaKarve glad to see your first contribution here!

jeff-phillips-18 commented:
Adding an action log for the link might be a better choice than adding a tab. Just my $00.02

jeff-phillips-18 commented on src/components/FieldLevelHelp/FieldLevelHelp.js:
```diff
@@ -0,0 +1,53 @@
+    >
+      <label>
+        {children + ' '}
+        <Icon type="pf" name={'info'} style={{ color: '#0088ce' }} />
```
we shouldn't use an inline style here add the `btn-link` class instead.

AparnaKarve replied:
`btn-link` introduces an underline that we typically see for links on a mouse hover, as shown below -
<img width="100" alt="screen shot 2018-01-24 at 11 29 32 am" src="https://user-images.githubusercontent.com/1538216/35352819-00a28a34-00fa-11e8-8632-d286bba67c9a.png">

Is there a way to remove that?

dabeng commented:
Hi @AparnaKarve , It will be better if you enable KNOBS panel to change "placement" or "trigger" properties 😊

ohadlevy commented:
and please add tests :-)

AparnaKarve commented:
@serenamarie125 

> #1 - can you change the placement of the popover so it "points" at the "I" icon as shown here in the design: http://www.patternfly.org/pattern-library/forms-and-controls/help-on-forms/#design

This is done. Can you please check the storybook again.

> #2 - per the design, I think we only support with popover, not tooltip - @jgiardino what do you think? should we remove the tooltip option to be inline with the documentation and ensure consistency?

Let me know what we decide. IMO, tooltips are nice to have.

#3 - this is a nit, but wondering if you could change the example to bring up a new tab when the link in the popover is clicked ? That would be the desired behavior, so if it's easy to change it may be worth it

Addressed the above.

AparnaKarve commented:
@jeff-phillips-18 

> Adding an action log for the link might be a better choice than adding a tab. Just my $00.02

The link is inside the popover. Not sure how to get the clicking action output in the action log.

AparnaKarve commented:
> It will be better if you enable KNOBS panel to change "placement" or "trigger" properties

@dabeng The knobs that are applicable in this case are `contentType` and `content` and the `Field Label` itself. I have added those. Can you review the [storybook](https://rawgit.com/AparnaKarve/patternfly-react/add_fieldlevelhelp_component-storybook/index.html?knob-contentType=popover&knob-content=Enter%20the%20hostname%20in%20a%20valid%20format%20%3Cbr%3E%20%20%3Ca%20target%3D%22_blank%22%20href%3D%22http%3A%2F%2Fwww.test.example.com%22%3EClick%20here%20for%20examples%20of%20valid%20hostnames%3C%2Fa%3E&knob-Field%20Label=Hostname&selectedKind=FieldLevelHelp&selectedStory=with%20Popover&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs) once again?

AparnaKarve commented:
@ohadlevy I have added some tests that pass in my environment.

Not sure what the below CI error means -
```
New snapshot was not written. The update flag must be explicitly passed to write a new snapshot.
    
    This is likely because this test is run in a continuous integration (CI) environment in which snapshots are not written by default.
```

Did I miss something?

AparnaKarve commented:
@priley86 @ohadlevy 

Tests in this PR seem to be failing because of the snapshot files.

In the travis configuration, do we need to delete the snapshot files prior to `npm run test`?

I managed to recreate the CI failure locally, and it looks like running 
`./node_modules/.bin/jest --updateSnapshot` before the tests execute, seems to fix it.

ohadlevy commented:
@AparnaKarve you simply need to include the snapshot file too

AparnaKarve commented:
Thanks @ohadlevy - that fixed it.

jeff-phillips-18 commented:
I would agree that this particular component only be onClick per the design at http://www.patternfly.org/pattern-library/forms-and-controls/help-on-forms/#design

Tooltips are useful, but I think we want this component to follow the design to maintain consistency across applications using this.

ohadlevy commented on src/components/FieldLevelHelp/FieldLevelHelp.js:
```diff
@@ -0,0 +1,58 @@
+
+  return (
+    <div>
+      <label>{children + ' '}</label>
```
nitpick: please write as: 
```js
<label>{`${children} `}</label>
```

AparnaKarve replied:
Addressed in 2dd035d6034b21b780d5fc031b497172730c9348

ohadlevy commented on src/components/FieldLevelHelp/FieldLevelHelp.js:
```diff
@@ -0,0 +1,58 @@
+    contentType === 'popover' ? (
+      <Popover id="{contentType}">{htmlContent}</Popover>
+    ) : (
+      <Tooltip id="{contentType}">{htmlContent}</Tooltip>
```
:+1: to removing tooltip here, its not part of the pf-design.

AparnaKarve replied:
Not removing the tooltip option in this iteration.
Keeping it temporarily for demo purposes, to explain an issue related to Formgroups and Popover below - https://github.com/patternfly/patternfly-react/pull/179#discussion_r164255291

ohadlevy commented on src/components/FieldLevelHelp/FieldLevelHelp.stories.js:
```diff
@@ -0,0 +1,47 @@
+import { defaultTemplate } from '../../../storybook/decorators/storyTemplates';
+import { DOCUMENTATION_URL } from '../../../storybook/constants';
+import { FieldLevelHelp } from './index';
+
```
maybe we should add it under the forms stories? it makes sense to me to see a complete form with field level help instead of a standalone?

AparnaKarve replied:
In 6c9f1d2, I have added the story under Forms.

It was a good call to add the FieldLevelHelp use case in the Form, since it looks like I may have stumbled upon a bug  in the Popover (or OverlayTrigger) component.

When the Popover component is used in conjunction with the `FormGroup` component, the Formgroup component steals the focus from Popover, causing the popover to close immediately. This happens in a fraction of a second, giving a visual impression that the popover never opened.

Please see a demo of the above issue here - https://aparnakarve.github.io/patternfly-react
under the Forms->Horizontal Form->Phone field

For this particular use case, a workaround would be to keep the Popover open - (An option that you requested here - https://github.com/patternfly/patternfly-react/pull/179#discussion_r163801743). That way, we would at least see the popover contents.
The 'Close Popover' knob demonstrates the above (set the value to false)

The other workaround is to use Tooltips - but since we are not considering tooltips for the FieldLevelHelp component, that is probably not an option.
The 'Popover/Tooltip' knob demonstrates the above (set the value to tooltip)

ohadlevy commented on src/components/FieldLevelHelp/FieldLevelHelp.test.js:
```diff
@@ -0,0 +1,30 @@
+    <FieldLevelHelp id="fieldlevelname1">Port Number</FieldLevelHelp>
+  );
+
+  let tree = component.toJSON();
```
I assume you can use const here?

AparnaKarve replied:
Addressed in 2dd035d6034b21b780d5fc031b497172730c9348

ohadlevy commented on src/components/FieldLevelHelp/FieldLevelHelp.js:
```diff
@@ -0,0 +1,58 @@
+  /** Contents displayed with popover or tooltip  */
+  content: PropTypes.string,
+  /** children nodes  */
+  children: PropTypes.node
```
does it make sense to have a props to force the popover to be open? (instead on just on hover/click)

AparnaKarve replied:
Added a new prop - `close`, to close/open the popover if the use case needs it.
(if popover is forced to be open, it can be closed after clicking on the 'i' icon)

0db70f4

jgiardino commented:
Thanks for adding this @AparnaKarve!

I also agree that this should only include the popover, and not the tooltip, to be consistent with the [patternfly design documentation](http://www.patternfly.org/pattern-library/forms-and-controls/help-on-forms/#design).

The only issue I see is with keyboard accessibility. I should be able to use the Tab key to place focus on the element that triggers the popover, and the Enter key to toggle the popover. Currently, the help icon is implemented as a `<span>` which can't receive keyboard focus. Alternatively, the [bootstrap example](https://getbootstrap.com/docs/3.3/javascript/#popovers) uses a `<button>` and the [patternfly example](http://www.patternfly.org/pattern-library/widgets/#popover) uses `<a href="#">`. (I think the `<button>` is more appropriate for this case, since we're not navigating anywhere.) 

Also, how does this component fit in with the [Forms component](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Disabled=false&knob-Show%20Help=true&knob-Show%20Modal=true&knob-Show%20Loading=false&knob-Show%20Labels=false&selectedKind=Forms&selectedStory=Horizontal%20Form&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs) that currently exists? For example, will I be able to use the Forms component and include this FieldLevelHelp component? Or is additional work needed for either component to support this? And should we include a storybook that shows an example that uses both components? (We can create separate issues to track this work)

AparnaKarve commented:
@jgiardino Thanks for the above feedback.

These are the most recent changes -
- removed tooltips
- implemented help icon as a button
(this seems to have resolved a focus issue that I was seeing earlier, mentioned in https://github.com/patternfly/patternfly-react/pull/179#discussion_r164255291, so thanks for the tip)
- Added FieldLevelHelp in the Form stories

Let me know if you find anything else. Thanks.

jgiardino commented on src/components/FieldLevelHelp/FieldLevelHelp.js:
```diff
@@ -1,63 +1,51 @@
+    >
+      <Button
+        bsStyle="link"
+        style={{ textDecoration: 'none', outline: 'none' }}
```
Can we remove these styles? Removing the outline results in an accessibility issue, because focus is not obvious to the sighted user who uses a keyboard to navigate. 

Also, as a general rule, if styles are ever needed, chances are they should be added to the core patternfly repo so that all JS repos can benefit from them. There are probably rare cases where react has a unique case that exists only in the context of the react repo. In that case we would add .less and .scss files for the component. I just now opened PR #195 to the [contributing guide](https://github.com/jgiardino/patternfly-react/blob/contributing/CONTRIBUTING.md#adding-styling-for-your-components) that includes some of these updates, in case you want to review and provide feedback :-)

jgiardino commented:
Yay, I'm glad using the button solved another issue. That was a happy accident!

I made a comment inline about inline styles. Related to this, I do realize that when the outline is visible, it's right up next to the text which is a little odd. But since there is no html implementation of this pattern in patternfly core that provides an example for us to follow, then I think for now what you have (after you remove the inline styles) is fine. I created an issue in patternfly to capture this: https://github.com/patternfly/patternfly/issues/943

Thank you for updating the Form examples to show the inline help. When I was looking at the html for this example, I noticed some weirdness. The elements `div.form-group` and `input.form-control` are both getting attributes `content` and `close`:
```
<div content="Please specify Country code <br> <a target='_blank' href='https://countrycode.org/'>Click here for a list of Country codes</a>" close="true" class="form-group">
  <label for="phone" class="col-sm-3 control-label">
    Phone
    <button type="button" class="btn btn-link" style="text-decoration: none; outline: none;">
      <span aria-hidden="true" class="pficon pficon-info"></span>
    </button>
  </label>
  <div class="col-sm-9">
    <input type="phone" content="Please specify Country code <br> <a target='_blank' href='https://countrycode.org/'>Click here for a list of Country codes</a>" close="true" id="phone" class="form-control">
    <span class="help-block">Enter a valid phone number</span>
  </div>
</div>
```

AparnaKarve commented:
@jgiardino I have removed the inline styling and have also addressed the html weirdness from the form stories https://github.com/patternfly/patternfly-react/pull/179#issuecomment-361762436
Please take another look. Thanks.

jeff-phillips-18 commented:
I know it's an issue in lots of places but the focus border on the info tip button encroaches on the label:
![image](https://user-images.githubusercontent.com/11633780/35623716-9282dcb4-065a-11e8-8caf-128c26be1cef.png)

Can we set less horizontal padding and more horizontal margin to pull in the focus border a bit?

sharvit commented:
@jeff-phillips-18, maybe it should just not implemented as a button? I think using a link, or just an icon can solve it without changing the styles (maybe just change the mouse pointer).

AparnaKarve commented:
> I think using a link, or just an icon can solve it without changing the styles (maybe just change the mouse pointer).

@sharvit My initial approach to this was just using an icon, but that had it's own set of issues. 
One of the major ones being, the one described here - https://github.com/patternfly/patternfly-react/pull/179#discussion_r164255291. This focus issue should still be viewable on my github.io fork here -
https://aparnakarve.github.io/patternfly-react
There are some knobs there to keep the popover open, which can be used as a workaround so that the Formgroup input does not steal focus from the open popover.

The other issue with the icon approach is what @jgiardino mentioned above, about keyboard accessibility.

jgiardino commented:
Using `<a>` instead of `<button>` has no effect on the outline issue. Using one of these is necessary due to keyboard accessibility.

Regarding @jeff-phillips-18 suggestion, I'm fine with adding css to the pf-react repo to address the outline issue until pf core has an implementation that we can use. I would suggest something like this:
`.field-help-pf: {outline-offset: -6px;}`

With the button getting the class: `<button class="btn btn-link pf-field-level" ...>` 

_I pinged the css-ers under the patternfly group on slack for feedback about this too, so I may edit this in the near future if I get an alternate suggestion from them._

AparnaKarve commented:
The css issue described above is outside the scope of this PR, and will be handled as a separate issue (https://github.com/patternfly/patternfly/issues/943)
That said, @jgiardino @jeff-phillips-18 Is there anything else that I need to address from my end to move this PR forward?

AparnaKarve commented on src/components/FieldLevelHelp/FieldLevelHelp.js:
```diff
@@ -0,0 +1,49 @@
+
+FieldLevelHelp.propTypes = {
+  /** additional fieldlevelhelp classes */
+  content: PropTypes.string,
```
@jeff-phillips-18 Addressed the above. Please review e720cf7

Moved the `dangerouslySetInnerHTML` usage in the stories (just like the other component stories are doing it)

AparnaKarve commented:
@jeff-phillips-18 LMK if there is anything else that I can address.
Thanks.

priley86 commented on src/components/Form/Stories/BasicForm.js:
```diff
@@ -90,5 +102,10 @@ export const getBasicFormKnobs = () => ({
-  disabled: boolean('Disabled', false)
+  disabled: boolean('Disabled', false),
+  content: text(
+    'FieldLevelHelp Content',
```
minor nit here: `Field Level Help Content`

AparnaKarve replied:
Addressed in f5628db87128706c1c309208f877c606cf4ae74f

priley86 commented on src/components/Form/Stories/SupportedControlsForm.js:
```diff
@@ -89,6 +90,17 @@ export const SupportedControlsFormFields = [
   },
+  {
+    controlId: 'FieldLevelHelp',
+    label: 'FieldLevelHelp',
```
same here: `Field Level Help`

AparnaKarve replied:
Addressed in f5628db87128706c1c309208f877c606cf4ae74f

priley86 commented:
other than two grammar nits this looks great to me, thanks @AparnaKarve !

priley86 commented:
one other thing i just now saw... your first commit needs to have commitizen syntax... so something like: 
```
`feat(field-level-help): adds field level help to forms`
```

AparnaKarve commented:
@priley86 I have reworded the first commit as you have indicated above.

priley86 commented:
great job @AparnaKarve ... this looks great to me.

priley86 commented:
woohoo 100% @AparnaKarve ! 🎉 🎉 thanks!

AparnaKarve commented:
Thanks everyone for your valuable feedback!
